### PR TITLE
Fixed bug on histogram count and add 1D+2D type for dqmio files

### PR DESCRIPTION
**Pull request by vicha-w**
Hi,

I have fixed a bug where a DQMIO file is shown with incorrect histogram count when 1D histograms are parsed twice. Since DQMIO files can contain 1D+2D histograms together, I have added a new type of dimensionality called "1D+2D".

As always, please let me know what you think.

Regards,
Vichayanun

### CODE REVIEW

File: histograms/models.py [hunk 1]
```diff
@@ -453,7 +457,11 @@ def from_nanodqm(file_path, data_era: str = "", read_chunk_lumi: int = -1):
         histogram_data_file, created = HistogramDataFile.objects.get_or_create(
             filepath=file_path
         )
-        histogram_data_file.data_dimensionality = HistogramDataFile.DIMENSIONALITY_2D
+        #histogram_data_file.data_dimensionality = HistogramDataFile.DIMENSIONALITY_2D
+        if histogram_data_file.data_dimensionality == HistogramDataFile.DIMENSIONALITY_UNKNOWN:
+            histogram_data_file.data_dimensionality = HistogramDataFile.DIMENSIONALITY_2D
+        elif histogram_data_file.data_dimensionality == HistogramDataFile.DIMENSIONALITY_1D:
+            histogram_data_file.data_dimensionality = HistogramDataFile.DIMENSIONALITY_1D2D
```
Comment (nothingface0) [inline histograms/models.py:464]: I suppose `1D2D` is not intended, right?

File: histograms/models.py [hunk 2]
```diff
@@ -211,7 +211,10 @@ def from_nanodqm(file_path, data_era: str = ""):
         histogram_data_file, created = HistogramDataFile.objects.get_or_create(
             filepath=file_path
         )
-        histogram_data_file.data_dimensionality = HistogramDataFile.DIMENSIONALITY_1D
+        if histogram_data_file.data_dimensionality == HistogramDataFile.DIMENSIONALITY_UNKNOWN:
+            histogram_data_file.data_dimensionality = HistogramDataFile.DIMENSIONALITY_1D
+        elif histogram_data_file.data_dimensionality == HistogramDataFile.DIMENSIONALITY_2D:
+            histogram_data_file.data_dimensionality = HistogramDataFile.DIMENSIONALITY_1D2D
```
Comment (nothingface0) [inline histograms/models.py:217]: Here too

File: histograms/models.py (hunk 1 above)
Comment (vicha-w) [reply histograms/models.py:464]: It is intended. This `1D2D` type should indicate that the file has 1D and 2D histograms parsed.

**Comment by nothingface0** [comment]
**Note:** I'm assuming that there is a possibility of having nanoDQMIO files with both Run and Lumisection Histograms. 

I understand what you want to do, but this could be a little confusing. The combination of the `data_dimensionality` and `granularity` fields were intended to represent the data contained in a file, which was assumed to be of one type. Having a combination of `data_dimensionality` values, a single value could make a programmer assume that this could apply to any data found in the file (even if the data is Run Histograms). We know this is not possible, but it's neither fool-proof or future-proof (you never know what type of data you will want to process some day).

Ideally, if we want to indicate that a file might contain multiple types of data, it's preferable to use a different approach:
- Create a new table, named something like `HistogramDataFileContents` and move there all attributes related to the dimensionality and granularity, i.e. with the following fields:
  ```python
  data_dimensionality = models.PositiveIntegerField(
        default=DIMENSIONALITY_UNKNOWN, choices=HISTOGRAM_DIMENSIONS_CHOICES, blank=True
    )
  granularity = models.CharField(
        max_length=3,
        choices=DATAFILE_GRANULARITY_CHOICES,
        default=GRANULARITY_UNKNOWN,
        help_text="The granularity of the data contained in the "
        "data file (either whole run or lumisections)",
    ) 
  ```
  This is due to the fact that, since we allow support for nanoDQMIO files now, a single `HistogramDataFile` might contain multiple types of data, right? `LumisectionHistograms1D`,  `LumisectionHistograms2D` and `RunHistograms` could, in principle, be stored in a single file.
- Remove the `data_dimensionality` and `granularity` fields from the `HistogramDataFile` model and replace them with a `ManyToManyField` which is of type `HistogramDataFileContents`:
  ```python
  contents = models.ManyToManyField(HistogramDataFileContents)
  ```
  This will now allow you to have multiple entries for a single file, i.e.: a single `HistogramDataFile` will be able to point to multiple entries in the `HistogramDataFileContents`.
  
This approach is preferable, because it can also scale with an increasing number of combinations. It is also future-proof because it can be extended more easily.

Would you like to try to implement it as I described above?

**Comment by vicha-w** [comment]
Thank you for your detailed solution. I'll try to implement them and let you know once it's finished.

**Review by nothingface0 — APPROVED** [review]
Left some cleanup comments. Once you have gone over them, please run the following:

```python
rm histogram_file_manager/migrations/0005_alter_histogramdatafile_data_dimensionality.py histogram_file_manager/migrations/0006_histogramdatafilecontents_and_more.py 
python manage.py migrate histogram_data_files 0004
python manage.py makemigrations histogram_data_files
```
This will remove the `1D+2D` migrations which you have already added to the PR which will no longer be needed.
It will also create a new `0005_....py` migration file which you will have to commit to this PR.

File: histogram_file_manager/models.py [hunk 3]
```diff
@@ -3,6 +3,38 @@
 
 # from django.conf import settings
 
+class HistogramDataFileContents(models.Model):
+    GRANULARITY_UNKNOWN = "unk"
```
Comment (nothingface0) [inline histogram_file_manager/models.py:7]: I _think_ we don't need `UNKNOWN` anymore. Could you try removing it completely?

File: histogram_file_manager/models.py [hunk 4]
```diff
@@ -3,6 +3,38 @@
...
+    GRANULARITY_RUN = "run"
+    GRANULARITY_LUMISECTION = "lum"
+    DIMENSIONALITY_UNKNOWN = 0
```
Comment (nothingface0) [inline histogram_file_manager/models.py:10]: Also here

File: histogram_file_manager/models.py [hunk 5]
```diff
@@ -3,6 +3,38 @@
...
+    DIMENSIONALITY_1D = 1
+    DIMENSIONALITY_2D = 2
+
+    HISTOGRAM_DIMENSIONS_CHOICES = (
+        (DIMENSIONALITY_UNKNOWN, "Unknown"),
```
Comment (nothingface0) [inline histogram_file_manager/models.py:15]: Also here

File: histogram_file_manager/models.py [hunk 6]
```diff
@@ -3,6 +3,38 @@
...
+        (DIMENSIONALITY_1D, "1D"),
+        (DIMENSIONALITY_2D, "2D"),
+    )
+
+    DATAFILE_GRANULARITY_CHOICES = (
+        (GRANULARITY_UNKNOWN, "Unknown"),
```
Comment (nothingface0) [inline histogram_file_manager/models.py:21]: Also here

File: histogram_file_manager/models.py [hunk 7]
```diff
@@ -3,6 +3,38 @@
...
+    data_dimensionality = models.PositiveIntegerField(
+        default = DIMENSIONALITY_UNKNOWN, 
+        choices = HISTOGRAM_DIMENSIONS_CHOICES, 
+        blank   = True
+    )
+
+    granularity = models.CharField(
+        max_length = 3,
+        choices    = DATAFILE_GRANULARITY_CHOICES,
+        default    = GRANULARITY_UNKNOWN,
```
Comment (nothingface0) [inline histogram_file_manager/models.py:35]: As above, make it `GRANULARITY_RUN` for now.

**Review by nothingface0 — CHANGES_REQUESTED** [review]

File: histogram_file_manager/models.py (hunk 3 above)
Comment (vicha-w) [reply histogram_file_manager/models.py]: Done.

File: histogram_file_manager/models.py (hunk 5 above)
Comment (vicha-w) [reply histogram_file_manager/models.py]: Done.

File: histogram_file_manager/models.py (hunk 4 above)
Comment (vicha-w) [reply histogram_file_manager/models.py]: Done.

File: histogram_file_manager/models.py (hunk 6 above)
Comment (vicha-w) [reply histogram_file_manager/models.py]: Done.

File: histogram_file_manager/models.py [hunk 8]
```diff
@@ -3,6 +3,38 @@
...
+    )
+
+    DATAFILE_GRANULARITY_CHOICES = (
+        (GRANULARITY_UNKNOWN, "Unknown"),
+        (GRANULARITY_RUN, "Run"),
+        (GRANULARITY_LUMISECTION, "Lumisection"),
+    )
+
+    data_dimensionality = models.PositiveIntegerField(
+        default = DIMENSIONALITY_UNKNOWN,
```
[the remark this answers is not in the document]
Comment (vicha-w) [reply histogram_file_manager/models.py]: Done.

File: histogram_file_manager/models.py (hunk 7 above)
Comment (vicha-w) [reply histogram_file_manager/models.py]: Done.

**Comment by vicha-w** [comment]
Hi. I have made some cleanup and migration in the latest commits as you requested. Unfortunately I also found a bug where 1D histograms can be imported twice into the database, so I wrote a check to see if an ME is already in the DB.

As always, please let me know if you would like further changes.

**Comment by nothingface0** [comment]
Looks good! Thank you Vicha.

**Pull request merged**